src/codegen/codegen_helper_visitor.cpp

```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
+ * three lists:
+ * - In NEURON/MOD2C implementation, we remove variables with NRNPRANGEIN
+ *   or NRNPRANGEOUT type
+ * - So who has NRNPRANGEIN and NRNPRANGEOUT type? these are such USEION read
+ *   or write variables that are not ionic currents.
+ * - This is the reason for mod files CaDynamics_E2.mod or cal_mig.mod, ica variable
+ *   is printed earlier in the list but other variables like cai, cao don't appear
+ *   in same order.
+ *
+ * Finally we create 4th list:
+ *  - variables with assigned property and not in the previous 3 lists
+ *
+ * We now print the variables in following order:
+ *
+ * - List 1 i.e. range + parameter variables are printed first
+ * - List 3 i.e. range + assigned variables are printed next
+ * - List 2 i.e. range + state variables are printed next
+ * - List 4 i.e. assigned and ion variables not present in the previous 3 lists
+ *
+ * NOTE:
+ * - State variables also have the property `assigned_definition` but these variables
+ *   are not from ASSIGNED block.
+ * - Variable can not be range as well as state, it's redeclaration error
+ * - Variable can be parameter as well as range. Without range, parameter
+ *   is considered as global variable i.e. one value for all instances.
+ * - If a variable is only defined as RANGE and not in assgined or parameter
+ *   or state block then it's not printed.
+ * - Note that if variable property is different than the variable type. For example,
+ *   if variable has range property, it doesn't mean the variable is declared as RANGE.
+ *   Other variables like STATE and ASSIGNED block variables also get range property.
+ * - Also, there is difference between declaration order vs. definition order. For
+ *   example, POINTER variable in NEURON block is just declaration and doesn't
+ *   determine the order in which they will get printed. Below we query symbol table
+ *   and order all instance variables into certain order.
  */
 void CodegenHelperVisitor::find_range_variables() {
     /// comparator to decide the order based on definition
     auto comparator = [](const SymbolType& first, const SymbolType& second) -> bool {
         return first->get_definition_order() < second->get_definition_order();
     };
 
+    /// from symbols vector `vars`, remove all ion variables which are not ionic currents
+    auto remove_non_ioncur_vars = [](SymbolVectorType& vars, CodegenInfo& info) -> void {
+        vars.erase(std::remove_if(vars.begin(),
+                                  vars.end(),
+                                  [&](SymbolType& s) {
+                                      return info.is_ion_variable(s->get_name()) &&
+                                             !info.is_ionic_current(s->get_name());
+                                  }),
+                   vars.end());
+    };
+
+    /// if `seconday` vector contains any symbol that exist in the `primary` then remove it
+    auto remove_var_exist = [](SymbolVectorType& primary, SymbolVectorType& secondary) -> void {
```

Comment:
another implementation is to stuff a hashtable with the values of primary and then run over the second and check if they are present in the hash table. This has complexity O(N_p + N_s) on average (N_p = primary.size(), N_s = primary.size()). Yours is O(N_p * N_s). However, hashing has a big constant in front so for small vectors your version could come on top. I would prefer the most O -complexity efficient since in case of small vectors the difference is not noticible anyway. Ultimately your decision